libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,304 @@
+		);
+	case evmasm::Instruction::LT:
+		return booleanValue(arguments.at(0) < arguments.at(1));
+	case evmasm::Instruction::SLT:
+		return booleanValue(unsignedToSigned(arguments.at(0)) < unsignedToSigned(arguments.at(1)));
+	case evmasm::Instruction::GT:
+		return booleanValue(arguments.at(0) > arguments.at(1));
+	case evmasm::Instruction::SGT:
+		return booleanValue(unsignedToSigned(arguments.at(0)) > unsignedToSigned(arguments.at(1)));
+	case evmasm::Instruction::EQ:
+		return booleanValue(arguments.at(0) == arguments.at(1));
+	case evmasm::Instruction::ISZERO:
+		return booleanValue(arguments.at(0) == constantValue(0));
+	case evmasm::Instruction::AND:
+		return smtutil::Expression::ite(
+			(arguments.at(0) == 0 || arguments.at(0) == 1) &&
+			(arguments.at(1) == 0 || arguments.at(1) == 1),
+			booleanValue(arguments.at(0) == 1 && arguments.at(1) == 1),
+			bv2int(int2bv(arguments.at(0)) & int2bv(arguments.at(1)))
+		);
+	case evmasm::Instruction::OR:
+		return smtutil::Expression::ite(
+			(arguments.at(0) == 0 || arguments.at(0) == 1) &&
+			(arguments.at(1) == 0 || arguments.at(1) == 1),
+			booleanValue(arguments.at(0) == 1 || arguments.at(1) == 1),
+			bv2int(int2bv(arguments.at(0)) | int2bv(arguments.at(1)))
+		);
+	case evmasm::Instruction::XOR:
+		return bv2int(int2bv(arguments.at(0)) ^ int2bv(arguments.at(1)));
+	case evmasm::Instruction::NOT:
+		return smtutil::Expression(u256(-1)) - arguments.at(0);
+	case evmasm::Instruction::SHL:
+		return smtutil::Expression::ite(
+			arguments.at(0) > 255,
+			constantValue(0),
+			bv2int(int2bv(arguments.at(1)) << int2bv(arguments.at(0)))
+		);
+	case evmasm::Instruction::SHR:
+		return smtutil::Expression::ite(
+			arguments.at(0) > 255,
+			constantValue(0),
+			bv2int(int2bv(arguments.at(1)) >> int2bv(arguments.at(0)))
+		);
+	case evmasm::Instruction::SAR:
+		return smtutil::Expression::ite(
+			arguments.at(0) > 255,
+			constantValue(0),
+			bv2int(smtutil::Expression::ashr(int2bv(arguments.at(1)), int2bv(arguments.at(0))))
+		);
+	// TODO ADDMOD
+	// TODO MULMOD
+	// TODO SIGNEXTEND
+	default:
+		break;
+	}
+	return newRestrictedVariable();
+}
+
+smtutil::Expression ReasoningBasedSimplifier::int2bv(smtutil::Expression _arg) const
+{
+	return smtutil::Expression::int2bv(std::move(_arg), 256);
+}
+
+smtutil::Expression ReasoningBasedSimplifier::bv2int(smtutil::Expression _arg) const
+{
+	return smtutil::Expression::bv2int(std::move(_arg));
+}
+
+smtutil::Expression ReasoningBasedSimplifier::newVariable()
+{
+	return m_solver->newVariable(uniqueName(), defaultSort());
+}
+
+smtutil::Expression ReasoningBasedSimplifier::newRestrictedVariable()
+{
+	smtutil::Expression var = newVariable();
+	m_solver->addAssertion(0 <= var && var < smtutil::Expression(bigint(1) << 256));
+	return var;
+}
+
+string ReasoningBasedSimplifier::uniqueName()
+{
+	return "expr_" + to_string(m_varCounter++);
+}
+
+shared_ptr<Sort> ReasoningBasedSimplifier::defaultSort() const
+{
+	return SortProvider::intSort();
```

Comment:
why not `uintSort`?

Reply:
Because of the wrapping? What happens if you write `0-1`?

Reply:
ah, yes :)